strapi/strapi
# [Data transfer] Filesystem agnostic loading of Strapi data files

### What does it do?

1. Improve error handling when schema or metadata cannot be loaded from a Strapi data file
2. Fix loading of Strapi data files that were generated on win32 systems
3. Strapi data files now always export with (internal) paths using posix format
4. Allows us to have jsonl directory paths deeper than one in our export

### Why is it needed?

1. When schema failed to load, it was incorrectly being reported as a schema validation error for every property on the schema
2. Exports that were generated on win32 had internal tar paths using backslash (\\) instead of forward slash (/) and could not be loaded
3. We want our generated tar files to be as consistent as possible between platforms, and it's better if we can know that our export files always use posix paths internally to reduce complexity in the future
4. it was a limitation we had previously that is no longer necessary with the new implementation

### How to test it?

Generate all four possible Strapi data files with a `yarn strapi export`:

- win32 with this PR
- win32 without this PR
- posix with this PR
- posix without this PR

Then test **importing all four of those files** on the same combinations:

- win32 with this PR 
- win32 without this PR
- posix with this PR
- posix without this PR

For each one:

- Ensure all assets are transferred and not just reported as successful with zero transfers
- Ensure all assets are transferred with the correct path and filename (ie, not with filenames that includes a path with slashes)
- Ensure all other data is transferred

### Known limitations

There are still some limitations that we can improve at a later time.

- if the importer does not read the assets path properly, it reports zero files transferred successfully rather than displaying an error for being unable to transfer any files. We need to add a sanity check in metadata about number of records for each stage (including entities, links, etc) so that we can detect when a different number of entities or assets are transferred than expected
- if a posix systems includes files in the assets directory that includes special characters not permitted on win32 (only possibly with manually placed files; strapi media library normalizes filenames to not include special characters) they will not unpack properly on win32; there is no solution for this, but it should be documented and we should check that it works with the check mentioned in the previous line

### Related issue(s)/PR(s)

Fixes #15616

Two community PRs were opened at roughly the same time to fix the issue, but only solved the issue for win32->win32 and did not solve the issue for win32->posix. I also needed to go a few steps further and improve the way we export the files to begin with, improved error handling, and cleaned up the inconsistent way we were comparing paths. Will add them as co-authors in a commit before publishing this PR :)

https://github.com/strapi/strapi/pull/16211
https://github.com/strapi/strapi/pull/16212
